File: python/coordinator.py
import logging
from select import select
from time import sleep
from os import getenv
from socket import socket, AF_INET, SOCK_DGRAM, SOCK_STREAM
from threading import Thread
from protobufs.pong_pb2 import Player, Query, Status, Match, Prepare, Message


MULTI_PONG_COORDINATOR_PORT = 4999
MULTI_PONG_SERVER_BUFFER = 512


class Coordinator:
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self.socket = socket(AF_INET, SOCK_STREAM)
        self.socket.bind(("0.0.0.0", MULTI_PONG_COORDINATOR_PORT))
        self.socket.listen(10)
        self.secret = getenv("MULTI_PONG_SECRET", "1")
        self.clients = [self.socket]
        self.searching_clients = []
        self.server_list = {
            ("127.0.0.1", 5000): -1,
            ("127.0.0.1", 5001): -1,
            ("127.0.0.1", 5002): -1,
            ("127.0.0.1", 5003): -1,
            ("127.0.0.1", 5004): -1,
        }

        Thread(target=self.check_status).start()
        self.listen()

    def _check_status(self):
        query_message = Message(query=Query())
        for server in self.server_list:
            self.send_message_to_server(server, query_message)

    def check_status(self):
        while True:
            self._check_status()
            self.matchmake()
            sleep(5)

    def send_message_to_server(self, server, message: Message):
        try:
            server_socket = socket(AF_INET, SOCK_DGRAM)
            server_socket.connect(server)
            server_socket.settimeout(1)
            server_socket.send(message.SerializeToString())

            received_data, _ = server_socket.recvfrom(MULTI_PONG_SERVER_BUFFER)

            message = Message()
            message.ParseFromString(received_data)
            content = message.WhichOneof("content")

            match content:
                case "status":
                    if message.status.phase == Status.Phase.WAITING:
                        self.logger.info("Server %s:%s is available", *server)
                    else:
                        self.logger.info("Server %s:%s is busy", *server)
                    self.server_list[server] = message.status.phase
                    return message.status
                case "tokens":
                    self.server_list[server] = Status.Phase.STARTED
                    return message.tokens
                case _:
                    self.logger.warning("Invalid message type %s from %s:%s", content, *server)

        except Exception:
            self.server_list[server] = -1
            self.logger.info("Server %s:%s is unresponsive", *server)

        finally:
            server_socket.close()

    def get_prepared_server(self):
        prepare_message = Message(prepare=Prepare(secret=self.secret))
        for server, status in self.server_list.items():
            if status != Status.Phase.WAITING:
                continue
            if tokens := self.send_message_to_server(server, prepare_message):
                return (server, tokens)
        return None

    def listen(self):
        self.logger.info("Started listening on %s:%s", "0.0.0.0", MULTI_PONG_COORDINATOR_PORT)

        while True:
            try:
                read_sockets, _, _ = select(self.clients, [], [])
                for socket in read_sockets:
                    if socket == self.socket:
                        sockfd, address = self.socket.accept()
                        self.clients.append(sockfd)
                        self.logger.info("Client %s:%s connected", *address)
                    else:
                        try:
                            if not (received_data := socket.recv(MULTI_PONG_SERVER_BUFFER)):
                                continue

                            message = Message()
                            message.ParseFromString(received_data)
                            content = message.WhichOneof("content")
                            
                            match content:
                                case "search":
                                    self.searching_clients.append(socket)
                                    self.logger.info("Add client %s:%s as a searching player", *socket.getpeername())
                                case _:
                                    self.logger.warning("Invalid message type %s from %s:%s", content, *socket.getpeername())
                        except:
                            socket.close()
                            self.clients.remove(socket)
                            if socket in self.searching_clients:
                                self.searching_clients.remove(socket)
                            self.logger.info("Client %s:%s disconnected", *socket.getpeername())
            except OSError as e:
                self.logger.debug("Socket exception - likely listening too early: %s", e)
            except Exception:
                self.logger.exception("Unhandled exception in client listen loop")


    def matchmake(self):
        if len(self.searching_clients) < 2:
            return

        if not (prepared_server := self.get_prepared_server()):
            return

        server, tokens = prepared_server

        for token, player_id in [(tokens.token_1, Player.Identifier.PLAYER_1), (tokens.token_2, Player.Identifier.PLAYER_2)]:
            player = Player(identifier=player_id, paddle_direction=0, paddle_location=0.5, score=0)
            match = Match(host=server[0], port=server[1], token=token, player=player)
            client = self.searching_clients.pop(0)
            self.send_message_to_client(client, match)
            self.logger.info("Forwarded match on %s:%s to client with token %s", *server, token)

    def send_message_to_client(self, client, data):
        message = Message()
        getattr(message, type(data).__name__.lower()).CopyFrom(data)
        client.send(message.SerializeToString())


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format='[%(asctime)s] - %(levelname)s: %(message)s')
    Coordinator()

File: cpp/coordinator.cpp
#include "coordinator.h"
#include "tools/logger.h"

#include <thread>

using namespace multi_pong;

Coordinator::Coordinator() {
#ifdef _WIN32
    WSADATA wsa_data;
    if (WSAStartup(MAKEWORD(2, 2), &wsa_data) != 0) {
        Logger::error("Failed to initialise Winsock");
        return;
    }
#endif

    std::thread status_thread(&Coordinator::check_status, this);
    status_thread.detach();

    coordinator_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (coordinator_socket < 0) {
        Logger::error("Failed to create socket");
        return;
    }

    sockaddr_in addr{};
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port = htons(MULTI_PONG_COORDINATOR_PORT);

    int enable = 1;
#ifdef _WIN32
    setsockopt(coordinator_socket, SOL_SOCKET, SO_REUSEADDR, (char*)&enable, sizeof(enable));
    setsockopt(coordinator_socket, SOL_SOCKET, SO_EXCLUSIVEADDRUSE, (char*)&enable, sizeof(enable));
#else
    setsockopt(coordinator_socket, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(enable));
#endif

    int bound = bind(coordinator_socket, (struct sockaddr*)&addr, sizeof(addr));
    if (bound < 0) {
        Logger::error("Failed to bind socket to port ", MULTI_PONG_COORDINATOR_PORT, ": ", bound);
        return;
    }

    Logger::debug("Socket bound successfully to port ", MULTI_PONG_COORDINATOR_PORT);
    listen_clients();
}

Coordinator::~Coordinator() {
    close_socket(coordinator_socket);
#ifdef _WIN32
    WSACleanup();
#endif
}

void Coordinator::check_status() {
    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(MULTI_PONG_SERVER_CHECK_INTERVAL));

		for (const auto& [server, status] : server_list) {
            Message message;
			message.mutable_query()->CopyFrom(Query());
            send_message_to_server(server, message);
        }

        matchmake();
    }
}

void Coordinator::matchmake() {
    if (searching_clients.size() < 2) {
        return;
    }

	std::pair<std::string, int> server;
	Tokens tokens;

    if (!get_prepared_server(server, tokens)) {
        return;
    }

    std::vector<std::pair<std::string, Player::Identifier>> token_pairs = {
        { tokens.token_1(), Player::Identifier::Player_Identifier_PLAYER_1 },
        { tokens.token_2(), Player::Identifier::Player_Identifier_PLAYER_2 }
    };

    for (auto& [token, player_id] : token_pairs) {
        Player player;
        player.set_identifier(player_id);
        player.set_paddle_direction(Direction::STOP);
        player.set_paddle_location(0.5f);
        player.set_score(0);

        Match match;
        match.set_host(server.first);
        match.set_port(server.second);
        match.set_token(token);
        match.mutable_player()->CopyFrom(player);

        socket_t client = searching_clients.front();
		searching_clients.pop_front();

		Message match_message = Message();
		match_message.mutable_match()->CopyFrom(match);

        send_message_to_client(client, match_message);

        Logger::info("Forwarded match on ", server.first, ":", server.second, " to client with token ", token);
    }
}

bool Coordinator::get_prepared_server(std::pair<std::string, int>& prepared_server, Tokens& tokens) {
    Message prepare_message = Message();
    Prepare prepare;

	prepare.set_secret(secret);
    prepare_message.mutable_prepare()->CopyFrom(prepare);

    for (const auto& [server, status] : server_list) {
        if (status != Status::Phase::Status_Phase_WAITING) {
            continue;
        }

        auto token_message = send_message_to_server(server, prepare_message);
        if (token_message && token_message->has_tokens()) {
            prepared_server = server;
			tokens = token_message->tokens();
            return true;
        }
    }
    return false;
}

std::optional<Message> Coordinator::send_message_to_server(std::pair<std::string, int> server, Message message) {
    socket_t server_socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (server_socket < 0) {
        return std::nullopt;
    }

    sockaddr_in address{};
    address.sin_family = AF_INET;
    address.sin_port = htons(server.second);
    inet_pton(AF_INET, server.first.c_str(), &address.sin_addr);

    int sent = sendto(server_socket, message.SerializeAsString().c_str(), message.ByteSizeLong(), 0, (sockaddr*)&address, sizeof(address));
    if (sent < 0) {
        Logger::error("Failed to send message to server ", server.first, ":", server.second);
        close_socket(server_socket);
        return std::nullopt;
    }

    char buffer[MULTI_PONG_SERVER_BUFFER];
    sockaddr_in source_address{};
    socklen_t source_address_len = sizeof(source_address);

#ifdef _WIN32
    constexpr DWORD timeout = MULTI_PONG_SERVER_CHECK_TIMEOUT * 1000;
#else
    struct timeval timeout;
    timeout.tv_sec = MULTI_PONG_SERVER_CHECK_TIMEOUT;
    timeout.tv_usec = 0;
#endif

    setsockopt(server_socket, SOL_SOCKET, SO_RCVTIMEO, (const char*)&timeout, sizeof(timeout));
    int received = recvfrom(server_socket, buffer, sizeof(buffer) - 1, 0, (sockaddr*)&source_address, &source_address_len);

    close_socket(server_socket);

    if (received < 0) {
        Logger::info("Server ", server.first, ":", server.second, " is unresponsive");
        return std::nullopt;
    }

    buffer[received] = '\0';

    Message received_message;
    if (!received_message.ParseFromArray(buffer, received)) {
        Logger::warning("Failed to process data into a protobuf message: ", buffer);
        return std::nullopt;
    }

    switch (received_message.content_case()) {
        case Message::kStatus:
            if (received_message.status().phase() == Status::Phase::Status_Phase_WAITING) {
                Logger::info("Server ", server.first, ":", server.second, " is available");
            }
            else {
                Logger::info("Server ", server.first, ":", server.second, " is busy");
            }
            server_list[server] = received_message.status().phase();
            return received_message;
        case Message::kTokens:
            server_list[server] = Status::Phase::Status_Phase_STARTED;
            return received_message;
        default:
            Logger::warning("Invalid message type ", received_message.content_case(), " from server ", server.first, ":", server.second);
            return received_message;
        };
}

void Coordinator::listen_clients() {
    listen(coordinator_socket, SOMAXCONN);
    Logger::info("Starting listening on 0.0.0.0:", MULTI_PONG_COORDINATOR_PORT);

    while (true) {
		fd_set read_fds;
		FD_ZERO(&read_fds);

		FD_SET(coordinator_socket, &read_fds);
        socket_t max_fd = coordinator_socket;

        for (auto client : clients) {
            FD_SET(client, &read_fds);
			if (client > max_fd) {
                max_fd = client;
            }
        }

        if (select(max_fd + 1, &read_fds, nullptr, nullptr, nullptr) < 0) {
            continue;
		}

        if (FD_ISSET(coordinator_socket, &read_fds)) {
            sockaddr_in client_addr{};
            socklen_t len = sizeof(client_addr);

#ifdef _WIN32
            SOCKET client_socket = accept(coordinator_socket, (sockaddr*)&client_addr, &len);
            if (client_socket == INVALID_SOCKET) {
                Logger::warning("Failed to accept client connection: ", WSAGetLastError());
                continue;
            }
#else
            socket_t client_socket = accept(coordinator_socket, (sockaddr*)&client_addr, &len);
            if (client_socket < 0) {
                Logger::warning("Failed to accept client connection: ", errno);
                continue;
            }
#endif

            clients.push_back(client_socket);

            Logger::info("Client ", inet_ntoa(client_addr.sin_addr), ":", ntohs(client_addr.sin_port), " connected");
        }

        for (auto it = clients.begin(); it != clients.end();) {
            socket_t client_socket = *it;

            if (!FD_ISSET(client_socket, &read_fds)) {
                ++it;
                continue;
            }

            char buffer[MULTI_PONG_SERVER_BUFFER];
            int bytes = recv(client_socket, buffer, sizeof(buffer), 0);

            sockaddr_in client_addr{};
            socklen_t len = sizeof(client_addr);
            getpeername(client_socket, (sockaddr*)&client_addr, &len);

            if (bytes <= 0) {
                Logger::info("Client ", inet_ntoa(client_addr.sin_addr), ":", ntohs(client_addr.sin_port), " disconnected");
                close_socket(client_socket);
                searching_clients.erase(std::remove(searching_clients.begin(), searching_clients.end(), client_socket), searching_clients.end());
                it = clients.erase(it);
                continue;
            }

            if (bytes < MULTI_PONG_SERVER_BUFFER) {
                buffer[bytes] = '\0';
            }

            Logger::debug("Received: ", buffer);

            Message message;
            if (!message.ParseFromArray(buffer, bytes)) {
				Logger::warning("Failed to process data into a protobuf message: ", buffer);
                it++;
                continue;
            }

            switch (message.content_case()) {
                case Message::kSearch:
                    Logger::info("Added client ", inet_ntoa(client_addr.sin_addr), ":", ntohs(client_addr.sin_port), " as a searching player");
                    searching_clients.push_back(client_socket);
                    break;
                default:
                    Logger::warning("Invalid message type ", message.content_case(), " from client ", inet_ntoa(client_addr.sin_addr), ":", ntohs(client_addr.sin_port));
                    break;
			};

            it++;
        }
    }
}

void Coordinator::send_message_to_client(socket_t client_socket, const Message& message) {
    send(client_socket, message.SerializeAsString().c_str(), message.ByteSizeLong(), 0);
}

// todo: move to a single main.cpp which takes --server, --coordinator and --client as command line args with conditional cmake builds
// int main() {
//     Coordinator coordinator;
//     return 0;
// }